cpp/src/arrow/engine/substrait/util.cc

```diff
@@ -0,0 +1,106 @@
+
+namespace arrow {
+
+namespace engine {
+
+Status SubstraitSinkConsumer::Consume(cp::ExecBatch batch) {
+  // Consume a batch of data
+  bool did_push = producer_.Push(batch);
+  if (!did_push) return Status::ExecutionError("Producer closed already");
+  return Status::OK();
+}
+
+Future<> SubstraitSinkConsumer::Finish() {
+  producer_.Push(IterationEnd<arrow::util::optional<cp::ExecBatch>>());
+  if (producer_.Close()) {
+    return Future<>::MakeFinished();
+  }
+  return Future<>::MakeFinished(
+      Status::ExecutionError("Error occurred in closing the batch producer"));
+}
+
+Status SubstraitExecutor::MakePlan() {
+  ARROW_ASSIGN_OR_RAISE(auto serialized_plan,
+                        engine::internal::SubstraitFromJSON("Plan", substrait_json_));
+
+  auto maybe_plan_json = engine::internal::SubstraitToJSON("Plan", *serialized_plan);
+  RETURN_NOT_OK(maybe_plan_json.status());
+
+  std::vector<std::shared_ptr<cp::SinkNodeConsumer>> consumers;
+  std::function<std::shared_ptr<cp::SinkNodeConsumer>()> consumer_factory = [&] {
+    // All batches produced by the plan will be fed into IgnoringConsumers:
+    consumers.emplace_back(new SubstraitSinkConsumer{generator_});
+    return consumers.back();
+  };
+
+  // Deserialize each relation tree in the substrait plan to an Arrow compute Declaration
+  ARROW_ASSIGN_OR_RAISE(declerations_,
+                        engine::DeserializePlan(*serialized_plan, consumer_factory));
+
+  // It's safe to drop the serialized plan; we don't leave references to its memory
+  serialized_plan.reset();
+
+  // Construct an empty plan (note: configure Function registry and ThreadPool here)
+  return Status::OK();
+}
+
+Result<std::shared_ptr<RecordBatchReader>> SubstraitExecutor::Execute() {
+  for (const cp::Declaration& decl : declerations_) {
+    RETURN_NOT_OK(decl.AddToPlan(plan_.get()).status());
+  }
+
+  ARROW_RETURN_NOT_OK(plan_->Validate());
+
+  ARROW_RETURN_NOT_OK(plan_->StartProducing());
+
+  std::shared_ptr<RecordBatchReader> sink_reader = cp::MakeGeneratorReader(
+      schema_, std::move(*generator_), exec_context_.memory_pool());
+  return sink_reader;
+}
+
+Status SubstraitExecutor::Finalize() {
+  ARROW_RETURN_NOT_OK(plan_->finished().status());
+  return Status::OK();
+}
+
+Result<std::shared_ptr<RecordBatchReader>> SubstraitExecutor::GetRecordBatchReader(
+    std::string& substrait_json, std::shared_ptr<arrow::Schema> schema) {
+  cp::ExecContext exec_context(arrow::default_memory_pool(),
+                               ::arrow::internal::GetCpuThreadPool());
+
+  arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+  ARROW_ASSIGN_OR_RAISE(auto plan, cp::ExecPlan::Make());
+
+  arrow::engine::SubstraitExecutor executor(substrait_json, &sink_gen, plan, schema,
+                                            exec_context);
+  RETURN_NOT_OK(executor.MakePlan());
+
+  ARROW_ASSIGN_OR_RAISE(auto sink_reader, executor.Execute());
+
+  RETURN_NOT_OK(executor.Finalize());
```

Comment:
I would frankly expect that the `RecordBatchReader`'s `Close` gets wired up to the executor's `Close` so that you don't have to manage their lifetimes independently, also, that would let you truly stream data right?

Reply:
What do you mean by the  `RecordBatchReader`'s Close?

Reply:
cc @lidavidm ping.

Reply:
ping @lidavidm I think converged on this.